Travis build passes, but a test fails

E.g, for build https://travis-ci.org/garris/BackstopJS/jobs/504019959 the result is 
`
14 passing (198ms)
  3 pending
  1 failing
`

But travis is green.

The failing test: `should work without an option param` (test/core/util/makeConfig_spec.js)

Randomly checking older builds, this has been there for several months.

Found the cause of the bug, the test exited with exit code 0.
If we comment out this line in https://github.com/garris/BackstopJS/blob/master/cli/index.js#L47
```
    // process.exit(code || exitCode);
```
and make a test to fail (since now tests are fixed).
Now the tests fail as intended with exit code 1.
```
$ npm run unit-test
...
  17 passing (223ms)
  3 pending
  1 failing

  1) the cli usage
       should print usage hints correctly:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert(false)

      + expected - actual

      -false
      +true
      
      at Context.<anonymous> (test/cli/usage_spec.js:8:5)



npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! backstopjs@4.3.4 unit-test: `mocha --reporter spec --recursive 'test/**/*_spec.js'`
npm ERR! Exit status 1
```

Not sure why we did `process.exit(code || exitCode);` or how to fix this.

Also, the related test is https://github.com/garris/BackstopJS/blob/master/test/cli/index_spec.js

--- a/cli/index.js
+++ b/cli/index.js
@@ -20,31 +20,22 @@ process.on('unhandledRejection', function (error) {
 
 if (argsOptions.h || argsOptions.help) {
   console.log(usage);
-  process.exit();
+  return;
 }
 
 if (argsOptions.v || argsOptions.version) {
   console.log('BackstopJS v' + version);
-  process.exit();
+  return;
 }
 
 var commandName = argsOptions['_'][0];
 
 if (!commandName) {
   console.log(usage);
-  process.exit();
 } else {
-  var exitCode = 0;
   console.log('BackstopJS v' + version);
   runner(commandName, argsOptions).catch(function () {
-    exitCode = 1;
-  });
-
-  /*
-   * Wait for the stdout buffer to drain.
-   */
-  process.on('exit', function (code) {
-    process.exit(code || exitCode);
+    process.exitCode = 1;
   });
 
   process.on('uncaughtException', function (err) {

--- a/cli/index.js
+++ b/cli/index.js
@@ -5,41 +5,45 @@ var usage = require('./usage');
 var version = require('../package.json').version;
 var runner = require('../core/runner');
 
-var argsOptions = parseArgs(process.argv.slice(2), {
-  boolean: ['h', 'help', 'v', 'version', 'i', 'docker'],
-  string: ['config'],
-  default: {
-    config: 'backstop.json'
+main();
+
+function main () {
+  var argsOptions = parseArgs(process.argv.slice(2), {
+    boolean: ['h', 'help', 'v', 'version', 'i', 'docker'],
+    string: ['config'],
+    default: {
+      config: 'backstop.json'
+    }
+  });
+
+  // Catch errors from failing promises
+  process.on('unhandledRejection', function (error) {
+    console.error(error && error.stack);
+  });
+
+  if (argsOptions.h || argsOptions.help) {
+    console.log(usage);
+    return;
   }
-});
 
-// Catch errors from failing promises
-process.on('unhandledRejection', function (error) {
-  console.error(error && error.stack);
-});
+  if (argsOptions.v || argsOptions.version) {
+    console.log('BackstopJS v' + version);
+    return;
+  }
 
-if (argsOptions.h || argsOptions.help) {
-  console.log(usage);
-  return;
-}
+  var commandName = argsOptions['_'][0];
 
-if (argsOptions.v || argsOptions.version) {
-  console.log('BackstopJS v' + version);
-  return;
-}
+  if (!commandName) {
+    console.log(usage);
+  } else {
+    console.log('BackstopJS v' + version);
+    runner(commandName, argsOptions).catch(function () {
+      process.exitCode = 1;
+    });
 
-var commandName = argsOptions['_'][0];
-
-if (!commandName) {
-  console.log(usage);
-} else {
-  console.log('BackstopJS v' + version);
-  runner(commandName, argsOptions).catch(function () {
-    process.exitCode = 1;
-  });
-
-  process.on('uncaughtException', function (err) {
-    console.log('Uncaught exception:', err.message, err.stack);
-    throw err;
-  });
+    process.on('uncaughtException', function (err) {
+      console.log('Uncaught exception:', err.message, err.stack);
+      throw err;
+    });
+  }
 }

--- a/test/cli/index_spec.js
+++ b/test/cli/index_spec.js
@@ -20,7 +20,7 @@ describe('cli', function () {
 
     require('../../cli/index');
 
-    promiseMock.finally(() => {
+    promiseMock.then(() => {
       assert.strictEqual(process.exitCode, undefined);
       assert(runnerMock.calledWith('test'));
       done();
@@ -35,7 +35,7 @@ describe('cli', function () {
 
     require('../../cli/index');
 
-    promiseMock.finally(() => {
+    promiseMock.catch(() => {
       assert.strictEqual(process.exitCode, 1);
       done();
     });

--- a/test/cli/index_spec.js
+++ b/test/cli/index_spec.js
@@ -1,23 +1,43 @@
-var mockery = require('mockery');
-var assert = require('assert');
+const mockery = require('mockery');
+const assert = require('assert');
+const sinon = require('sinon');
 
 describe('cli', function () {
-  before(function () {
-    mockery.enable({ warnOnUnregistered: false });
+  beforeEach(function () {
+    mockery.enable({ warnOnUnregistered: false, useCleanCache: true });
   });
 
-  after(function () {
+  afterEach(function () {
+    mockery.deregisterAll();
     mockery.disable();
   });
 
-  // causes issue with CI https://github.com/garris/BackstopJS/issues/1006
-  it.skip('should call the runner without custom options correctly', function (done) {
+  it('should call the runner without custom options correctly', function (done) {
     process.argv = ['node', 'backstop', 'test'];
-    mockery.registerMock('../core/runner', function (command, options) {
-      assert.strictEqual(command, 'test');
-      done();
-      return Promise.resolve();
-    });
+    const promiseMock = Promise.resolve();
+    const runnerMock = sinon.stub().returns(promiseMock);
+    mockery.registerMock('../core/runner', runnerMock);
+
     require('../../cli/index');
+
+    promiseMock.finally(() => {
+      assert.strictEqual(process.exitCode, undefined);
+      assert(runnerMock.calledWith('test'));
+      done();
+    });
+  });
+
+  it('should exit with code 1 if runner fails', function (done) {
+    process.argv = ['node', 'backstop', 'test'];
+    const promiseMock = Promise.reject('errorMock');
+    const runnerMock = sinon.stub().returns(promiseMock);
+    mockery.registerMock('../core/runner', runnerMock);
+
+    require('../../cli/index');
+
+    promiseMock.finally(() => {
+      assert.strictEqual(process.exitCode, 1);
+      done();
+    });
   });
 });

--- a/test/cli/index_spec.js
+++ b/test/cli/index_spec.js
@@ -29,7 +29,7 @@ describe('cli', function () {
 
   it('should exit with code 1 if runner fails', function (done) {
     process.argv = ['node', 'backstop', 'test'];
-    const promiseMock = Promise.reject('errorMock');
+    const promiseMock = Promise.reject(new Error('errorMock'));
     const runnerMock = sinon.stub().returns(promiseMock);
     mockery.registerMock('../core/runner', runnerMock);
 

